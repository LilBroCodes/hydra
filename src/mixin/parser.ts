import fs from 'fs';
import path from 'path';

const INJECT_REGEX = /\/\/\s*@Inject\s*\(\s*method\s*=\s*"([^"]+)"\s*,\s*at\s*=\s*"([^"]+)"(?:\s*,\s*ordinal\s*=\s*(\d+))?\s*\)/;

export function parse(code: string) {
    const lines = code.split('\n');

    const results: Array<{ match: { method: string; at: string; ordinal: number }; functionCode: string }> = [];

    for (let i = 0; i < lines.length; i++) {
        const match = INJECT_REGEX.exec(lines[i]);
        if (match) {
            let j = i + 1;

            while (j < lines.length && /^\s*(\/\/.*)?$/.test(lines[j])) {
                j++;
            }

            let braceCount = 0;
            let funcLines: string[] = [];
            let started = false;

            while (j < lines.length) {
                const line = lines[j];
                if (!started) {
                    if (/function\s+[\w$]+\s*\(.*\)\s*\{/.test(line) || /const\s+[\w$]+\s*=\s*\(?.*?\)?\s*=>\s*\{/.test(line)) {
                        started = true;
                    } else {
                        j++;
                        continue;
                    }
                }

                funcLines.push(line);
                braceCount += (line.match(/{/g) || []).length;
                braceCount -= (line.match(/}/g) || []).length;

                if (started && braceCount === 0) {
                    break;
                }

                j++;
            }

            results.push({
                match: {
                    method: match[1],
                    at: match[2],
                    ordinal: parseInt(match[3]) || 0,
                },
                functionCode: funcLines.join('\n'),
            });

            i = j;
        }
    }

    return results;
}

export function parseMixin(filename: string) {
    return parse(fs.readFileSync(path.resolve(filename), 'utf-8'));
}
