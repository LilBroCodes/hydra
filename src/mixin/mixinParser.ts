import fs from 'fs';
import { parse } from '@babel/parser';
import traverse from '@babel/traverse';
import generate from '@babel/generator';

export function parseMixinFile(filePath: string): { line: number, comment: string, definitionCode: string; }[] {
    return parseMixinCode(fs.readFileSync(filePath, 'utf8'));
}

export function parseMixinCode(code: string): { line: number, comment: string, definitionCode: string; }[] {
    const ast = parse(code, {
        sourceType: 'module',
        plugins: ['jsx', 'typescript'],
        attachComment: true,
        tokens: true,
        ranges: true,
        allowReturnOutsideFunction: true,
        allowAwaitOutsideFunction: true,
        errorRecovery: true,
    });

    const results: {
        line: number;
        comment: string;
        definitionCode: string;
    }[] = [];

    traverse(ast, {
        enter(path) {
            if (
                path.isFunctionDeclaration() ||
                path.isFunctionExpression() ||
                path.isArrowFunctionExpression() ||
                path.isClassDeclaration() ||
                path.isClassExpression()
            ) {
                const node = path.node;

                const leadingComments = (node.leadingComments || []).filter(c =>
                    c.type === 'CommentLine' && c.value.includes('@')
                );

                for (const comment of leadingComments) {
                    if (comment.loc && node.loc) {
                        if (comment.loc.end.line === node.loc.start.line - 1) {
                            const generated = generate(node, {}, code).code;
                            results.push({
                                line: comment.loc.start.line,
                                comment: comment.value.trim(),
                                definitionCode: generated,
                            });
                        }
                    }
                }
            }
        },
    });

    return results;
}
