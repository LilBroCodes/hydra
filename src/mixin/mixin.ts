import {Tag} from "./tags/tag";
import {parseMixinCode} from "./mixinParser";
import {BaseAt, ExtendedAt} from "./tags/at";
import {Inject} from "./tags/inject";
import {ModifyReturnValue} from "./tags/modifyReturnValue";
import fs from "fs";
import path from "path";

export class Mixin {
    code: string;
    methods: TaggedMethod[];

    constructor(code: string) {
        this.code = code;

        const rawMethods = parseMixinCode(this.code);
        this.methods = [];

        for (const method of rawMethods) {
            this.methods.push(new TaggedMethod(
                method.line,
                method.definitionCode,
                method.comment,
                Tag.fromString(method.comment)
            ));
        }
    }

    static ofFile(file: string) {
        return new Mixin(fs.readFileSync(path.resolve(file), "utf-8"));
    }
}

export class TaggedMethod {
    line: number;
    code: string
    comment: string
    tag: BaseAt | ExtendedAt | Inject | ModifyReturnValue | null

    constructor(line: number, code: string, comment: string, tag: BaseAt | ExtendedAt | Inject | ModifyReturnValue | null) {
        this.line = line;
        this.code = code;
        this.comment = comment;
        this.tag = tag;
    }
}