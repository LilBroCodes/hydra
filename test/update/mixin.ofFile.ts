import {Mixin} from "../../src/mixin/mixin";
import fs from "fs";
import path from "path";
import {TestUpdate} from "./updater";

export class MixinOfFile extends TestUpdate {
    runUpdate() {
        const mixin = Mixin.ofFile("test/resources/template.js");
        const output = JSON.stringify(mixin, null, 2);

        const outputPath = path.resolve("test/expected/mixin.ofFile.json");
        fs.writeFileSync(outputPath, output, "utf-8");
    }
}
