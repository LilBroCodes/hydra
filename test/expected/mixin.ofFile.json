{
  "lines": [
    "function unrelatedFunction() {",
    "    console.log(\"This is not an inject function.\");",
    "}",
    "",
    "// @Inject(method = \"onInit\", at = \"HEAD\", ordinal = 0)",
    "function initHook() {",
    "    console.log(\"Injected at onInit HEAD with ordinal 0\");",
    "}",
    "",
    "// @Inject(method=\"onRender\", at=\"TAIL\")",
    "const renderHook = () => {",
    "    console.log(\"Injected at onRender TAIL with no ordinal\");",
    "}",
    "const x = 42;",
    "",
    "// @Inject(method=\"onClick\", at=\"HEAD\", ordinal=2)",
    "function clickHook() {",
    "    const now = Date.now();",
    "    console.log(\"Click hook executed at\", now);",
    "}",
    "",
    "// @Inject(method = \"run(args)\", at=\"HEAD\")",
    "function runHook(args) {",
    "",
    "}",
    ""
  ],
  "code": "function unrelatedFunction() {\n    console.log(\"This is not an inject function.\");\n}\n\n// @Inject(method = \"onInit\", at = \"HEAD\", ordinal = 0)\nfunction initHook() {\n    console.log(\"Injected at onInit HEAD with ordinal 0\");\n}\n\n// @Inject(method=\"onRender\", at=\"TAIL\")\nconst renderHook = () => {\n    console.log(\"Injected at onRender TAIL with no ordinal\");\n}\nconst x = 42;\n\n// @Inject(method=\"onClick\", at=\"HEAD\", ordinal=2)\nfunction clickHook() {\n    const now = Date.now();\n    console.log(\"Click hook executed at\", now);\n}\n\n// @Inject(method = \"run(args)\", at=\"HEAD\")\nfunction runHook(args) {\n\n}\n",
  "methods": [
    {
      "match": {
        "method": "onInit",
        "at": "HEAD",
        "ordinal": 0
      },
      "code": "function initHook() {\n    console.log(\"Injected at onInit HEAD with ordinal 0\");\n}"
    },
    {
      "match": {
        "method": "onRender",
        "at": "TAIL",
        "ordinal": 0
      },
      "code": "const renderHook = () => {\n    console.log(\"Injected at onRender TAIL with no ordinal\");\n}"
    },
    {
      "match": {
        "method": "onClick",
        "at": "HEAD",
        "ordinal": 2
      },
      "code": "function clickHook() {\n    const now = Date.now();\n    console.log(\"Click hook executed at\", now);\n}"
    },
    {
      "match": {
        "method": "run(args)",
        "at": "HEAD",
        "ordinal": 0
      },
      "code": "function runHook(args) {\n\n}"
    }
  ]
}